### ISSUE TITLE
SocketConnection / Pipelines usage advice

### ISSUE THREAD

**Issue by galvesribeiro** [mgravell/Pipelines.Sockets.Unofficial#17]
Hello!

Trying to experiment pipelines and ended up on your blog and this package.

I have a case here where I'm experimenting with the library and pipelines for a socket client and for whatever reason the server is disconnecting me with read timeouts. This is basically the test class I'm using:

```cs
 public class PipeTestConnection
    {
        private const int TIMEOUT = 20000;
        private readonly string _endpoint;
        private readonly int _port;
        private SocketConnection _connection;

        public bool Connected => this._connection?.Socket != null && this._connection.Socket.Connected;

        public PipeTestConnection(string endpoint, int port)
        {
            this._endpoint = endpoint;
            this._port = port;
        }

        public async Task Connect()
        {
            var timeout = Task.Delay(TIMEOUT);
            var connect = SocketConnection.ConnectAsync(
                new IPEndPoint((Dns.GetHostAddresses(this._endpoint))[0], this._port),
                connectionOptions: SocketConnectionOptions.None);
            var first = await Task.WhenAny(timeout, connect);
            if (first == timeout) throw new TimeoutException();
            this._connection = connect.Result;
        }

        public async Task<byte[]> SendPacket(GPRequest packet)
        {
            if (packet == null) throw new ArgumentNullException(nameof(packet));
            if (!this.Connected) throw new InvalidOperationException("Not connected");

            var requestBuffer = packet.ToBuffer();

            await this._connection.Output.WriteAsync(requestBuffer).ConfigureAwait(false);
            var cancellationSource = new CancellationTokenSource();
            var timeout = Task.Delay(TIMEOUT);
            var receive = ReceivePacket(cancellationSource.Token);

            var first = await Task.WhenAny(timeout, receive).ConfigureAwait(false);
            if (first == timeout)
            {
                cancellationSource.Cancel();
                throw new TimeoutException();
            }
            return receive.Result;
        }

        private async Task<byte[]> ReceivePacket(CancellationToken token)
        {
            var readResult = await this._connection.Input.ReadAsync(token).ConfigureAwait(false);

            var buffer = readResult.Buffer;
            var responseBuffer = buffer.ToArray().Skip(2); // Theoretically I can discard the 2 bytes for the message size header when reading it as I'm only interested in the data.
            this._connection.Input.AdvanceTo(buffer.End);

            return responseBuffer.ToArray();
        }

        public void Disconnect() => this._connection.Dispose();
    }
```

The protocol the server expects is a 2 bytes in the beginning of the message with the content size and then the data following it. It is basically request/response based so the server will never send something to the client other than replying to its requests.

Sometimes server (which I don't control) disconnect me with read timeout and sometimes incomplete data arrive on the client and is returned by the `ReceivePacket` method.

Can you try pinpoint where I'm being dumb?

Thanks! Awesome blog post btw!

**Comment by mgravell** [comment]
is it genuinely a "read timeout", or is it the `TimeoutException` from `SendPacket` that we're talking about?

re " incomplete data arrive on the client and is returned by the ReceivePacket method." - yes, incomplete data is perfectly normal for network code - almost the default; and - since this is a TCP stream, we're not really talking packets here - just a stream; but... what I would *expect* would be something like:
```c#
private async Task<byte[]> ReceivePacket(CancellationToken token)
{
    while(true)
    {
            var readResult = await this._connection.Input.ReadAsync(token).ConfigureAwait(false);
            int payloadLen;
            var buffer = readResult.Buffer;
            if (buffer.Length < 2 // not enough data for the prefix
                  || readResult.Length < 2 + (payloadLen = ParseLengthPrefix(readResult.Buffer)))
            {
                // not enough data (yet)
                this._connection.Input.AdvanceTo(buffer.Start); // don't consume
                continue; // keep reading
            }
            var payload = buffer.Slice(2, payloadLength); // skip the prefix
            var responseBuffer = payload.ToArray()
            this._connection.Input.AdvanceTo(payload.End);

            return responseBuffer.ToArray(); // note: personally I'd use a right-sized leased chunk here, not a new array
        }
}
```

What this does is loop until we a: have at least 2 bytes for the prefix, b: have the payload content - then we extract the payload portion.

If you want guidance on the "right-sized leased chunk" comment, please see my blog here: https://blog.marcgravell.com/2018/07/pipe-dreams-part-3.html (and the 3.1 addendum)

**Comment by galvesribeiro** [comment]
Thanks for the quick reply @mgravell 

> is it genuinely a "read timeout", or is it the TimeoutException from SendPacket that we're talking about?
<img width="868" alt="image" src="https://user-images.githubusercontent.com/4714040/47189155-960ae000-d311-11e8-9184-d363522919ef.png">

Yeah, the timeout on the client eventually throw, but that is my code. When I refer to read timeout is because the server is showing that like in the picture. 

Reg. the incomplete message, I know that it is normal on networking, I was just wondering how to do that using those new APIs. 

So I tried out your code and I was able to read the full message properly. However, the timeouts still happen...

And about the leased chunk I'll check on that part of the blog since yes, I want to avoid the allocation there as well but first I need to understand why the timeouts happen on the server when sending message to the client...
